File: platforms/android/feedback/src/main/kotlin/io/markerusa/feedback/MultipartUploader.kt
package io.markerusa.feedback

import java.io.BufferedOutputStream
import java.net.HttpURLConnection
import java.net.URL

/**
 * Hand-rolled multipart/form-data upload.
 *
 * The ingest API takes multipart because browsers send FormData; matching that on
 * native keeps one endpoint rather than two. Without an HTTP library this means
 * writing the envelope by hand — tedious but self-contained, and it keeps the
 * dependency count at zero.
 */
internal class MultipartUploader(
    endpoint: String,
    private val projectId: String,
    private val timeoutMs: Int = 30_000
) {
    private val endpoint = endpoint.trimEnd('/')

    /** Returns the server-assigned attachment id. */
    fun upload(bytes: ByteArray, filename: String): String {
        val boundary = "----markerio${System.nanoTime()}"
        var connection: HttpURLConnection? = null

        try {
            connection = (URL("$endpoint/v1/uploads").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = timeoutMs
                readTimeout = timeoutMs
                doOutput = true
                setRequestProperty("content-type", "multipart/form-data; boundary=$boundary")
                setRequestProperty("x-project-id", projectId)
                // Screenshots can be megabytes; streaming avoids buffering the whole
                // request in memory on a device that may be low on it.
                setChunkedStreamingMode(16 * 1024)
            }

            BufferedOutputStream(connection.outputStream).use { out ->
                out.write(
                    ("--$boundary\r\n" +
                        "Content-Disposition: form-data; name=\"file\"; filename=\"$filename\"\r\n" +
                        "Content-Type: image/png\r\n\r\n").toByteArray(Charsets.UTF_8)
                )
                out.write(bytes)
                out.write(
                    ("\r\n--$boundary\r\n" +
                        "Content-Disposition: form-data; name=\"kind\"\r\n\r\n" +
                        "screenshot\r\n" +
                        "--$boundary--\r\n").toByteArray(Charsets.UTF_8)
                )
                out.flush()
            }

            val status = connection.responseCode
            if (status !in 200..299) {
                connection.errorStream?.use { it.readBytes() }
                throw IngestException("Upload failed with $status", status, status >= 500)
            }

            val body = connection.inputStream.use { it.readBytes() }.toString(Charsets.UTF_8)
            return extractId(body)
                ?: throw IngestException("Upload response missing id", status, false)
        } catch (e: IngestException) {
            throw e
        } catch (e: Exception) {
            throw IngestException(e.message ?: "Upload error", null, true)
        } finally {
            connection?.disconnect()
        }
    }

    /**
     * Pulls `id` out of `{"id":"..."}` without a JSON parser.
     *
     * Narrow by design: this reads one known field from our own server's response,
     * not arbitrary JSON. Anything more complex belongs in a real parser.
     */
    private fun extractId(body: String): String? {
        val marker = "\"id\""
        val keyIndex = body.indexOf(marker)
        if (keyIndex < 0) return null

        val colon = body.indexOf(':', keyIndex + marker.length)
        if (colon < 0) return null

        val open = body.indexOf('"', colon)
        if (open < 0) return null

        val close = body.indexOf('"', open + 1)
        if (close < 0) return null

        return body.substring(open + 1, close).ifEmpty { null }
    }
}

File: platforms/ios/Sources/SusaSignals/MultipartUploader.swift
import Foundation

/**
 multipart/form-data upload.

 The ingest API takes multipart because browsers send FormData; matching it here
 keeps one endpoint rather than two.
 */
final class MultipartUploader {
    private let endpoint: String
    private let projectId: String
    private let timeout: TimeInterval

    init(endpoint: String, projectId: String, timeout: TimeInterval = 30) {
        var trimmed = endpoint
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        self.endpoint = trimmed
        self.projectId = projectId
        self.timeout = timeout
    }

    /// Returns the server-assigned attachment id.
    ///
    /// `mimeType` is explicit because replay frames are JPEG while screenshots
    /// are PNG, and the server validates the declared type against an allowlist.
    func upload(pngData: Data, filename: String, mimeType: String = "image/png") throws -> String {
        guard let url = URL(string: endpoint + "/v1/uploads") else {
            throw IngestError(message: "Invalid endpoint", status: nil, retryable: false)
        }

        let boundary = "----susa-signals\(UInt64(Date().timeIntervalSince1970 * 1_000_000))"
        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "content-type")
        request.setValue(projectId, forHTTPHeaderField: "x-project-id")

        var body = Data()
        func append(_ string: String) {
            if let data = string.data(using: .utf8) { body.append(data) }
        }

        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n")
        append("Content-Type: image/png\r\n\r\n")
        body.append(pngData)
        append("\r\n--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"kind\"\r\n\r\n")
        append("screenshot\r\n")
        append("--\(boundary)--\r\n")

        request.httpBody = body

        let semaphore = DispatchSemaphore(value: 0)
        var responseData: Data?
        var response: URLResponse?
        var transportError: Error?

        URLSession.shared.dataTask(with: request) { data, res, err in
            responseData = data
            response = res
            transportError = err
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let transportError {
            throw IngestError(message: transportError.localizedDescription, status: nil, retryable: true)
        }

        guard let http = response as? HTTPURLResponse else {
            throw IngestError(message: "No HTTP response", status: nil, retryable: true)
        }

        guard (200...299).contains(http.statusCode) else {
            throw IngestError(
                message: "Upload failed with \(http.statusCode)",
                status: http.statusCode,
                retryable: http.statusCode >= 500
            )
        }

        guard
            let responseData,
            let object = try? JSONSerialization.jsonObject(with: responseData) as? [String: Any],
            let id = object["id"] as? String,
            !id.isEmpty
        else {
            throw IngestError(message: "Upload response missing id", status: http.statusCode, retryable: false)
        }

        return id
    }
}
